SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,435 @@
+    default:
+      return false
+    }
+  }
+  
+  func isDistinct(from other: AccessBase) -> Bool {
+    fatalError()
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+  
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+  
+  func isDistinct(from: AccessPath) -> Bool {
+    fatalError("TODO")
+  }
+}
+
+/// An `AccessRoot` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessRoot {
+  var root: Value
+  
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  var path: SmallProjectionPath
+}
+
+typealias AccessStorage = (maybeRoot: AccessRoot?, base: AccessBase)
+
+/// An `AccessScope` denotes a pair of `begin_access` `end_access`
+/// instructions.
+struct AccessScope {
+  // TODO: add support for begin_access_unpaired
+  var beginAccess: BeginAccessInst
+
+  var accessedAddress: Value { beginAccess.operand }
+
+  var accessKind: AccessKind { beginAccess.accessKind }
+  
+  var scopeEndingInstructions: EndAccessInstructions {
+    beginAccess.endAccesses
+  }
+}
+
+enum EnclosingScope {
+  case scope(AccessScope)
+  case base(AccessBase)
+}
+
+
+func canBeOperandOfIndexAddr(_ value: Value) -> Bool {
+  switch value {
+  case is IndexAddrInst, is PointerToAddressInst, is RefTailAddrInst:
+    return true
+  default:
+    return false
+  }
+}
+
+
+protocol HasMerge {
+  init()
+  mutating func merge(with: Self) -> Self?
+}
+
+struct CacheEntry<State : HasMerge> {
+  typealias Path = SmallProjectionPath
+  var valid: Bool = false
+  var state: State = State()
+  var path: Path = Path()
+  
+  mutating func needWalk(path: Path, state: State) -> CacheEntry? {
+    if !valid {
+      valid = true
+      self.state = state
+      self.path = path
+      return self
+    }
+    let newPath = self.path.merge(with: path)
+    let newState = self.state.merge(with: state)
+    if newPath != self.path || newState != nil {
+      return self
+    }
+    
+    return nil
+  }
+}
+
+protocol WithDefaultCache where Self : ValueUseDefWalker, Self.State : HasMerge {
+  typealias DefaultCache = Dictionary<HashableValue, CacheEntry<State>>
+  var walkUpCache: DefaultCache { get set }
+}
+
+extension WithDefaultCache {
+  
+  mutating func shouldRecomputeUp(def: Value, path: Path, state: State) -> (Path, State)? {
+    if let entry = walkUpCache[def.hashable, default: CacheEntry()].needWalk(path: path, state: state) {
+      return (entry.path, entry.state)
+    }
+    return nil
+  }
+}
+
+fileprivate struct AccessPathWalkerInternal : AddressUseDefWalker, ValueUseDefWalker, WithDefaultCache {
```

Comment:
> AccessPathWalkerInternal : AddressUseDefWalker, ValueUseDefWalker

It's really unfortunate that AccessPathWalkerInternal conforms to ValueUseDefWalker. It pulls in all the implementation used to walk values, which doesn't normally make sense for computing an access path. There's no way to tell anymore which operations may part of an access path. This partially defeats the purpose of separating the address and value walkers.

I think this was done to allow sequences like:
```
address_to_pointer
struct
struct_extract
pointer_to_address
```

But the operations on the pointer value don't contribute to the address projection path.  And the set of allowable operations on pointer types are completely different from those for addresses or references (non-trivial types).

I originally thought that multiple walkers would be involved in a single SSA traversal and they would just share a context. That makes sense to me because the pointer and address walkers should neither share the same state nor share any implementation. The pointer walker doesn't have anything to do with the access base or access path. None of the operations overlap.

I would define a completely separate pointer walker. You can assert that the pointer path is empty when it calls back to the parent walker. The AccessPathWalker would not need to deal with any value walking stuff.

Reply:
I agree. I'm even not sure if this logic belongs to the accesspath utilities. It's kind of a best-effort and is more an optimization rather than a "structural property".
Maybe we do this tracking from `pointer_to_address` back to the matching `address_to_pointer` in a separate utility function wich can be called by clients on demand.

Reply:
> It's kind of a best-effort and is more an optimization rather than a "structural property".
Maybe we do this tracking from pointer_to_address back to the matching address_to_pointer in a separate utility function which can be called by clients on demand.

Exactly.

It would be great to have a strategy that correctness and verification never require looking past a pointer_to_address. Note that this means it's illegal for any SIL pass to insert a pointer_to_address.

For correctness, the optimizer needs to "see" all global accesses though. Unfortunately, I think we run into cases where a global variable addressor can be cloned, resulting in a RawPointer-type phi. Then you need to both look past `pointer_to_address` and see that all pointers come from the same addressor to tell that it's a global variable access.

It would be better to handle this by defining a structural requirement for global addressors and preventing them from being cloned without cloning their uses (we do this sort of thing elsewhere). Maybe even using global_pointer_to_address instead of pointer_to_address.

If that makes sense, I'll file a radar so we don't forget.